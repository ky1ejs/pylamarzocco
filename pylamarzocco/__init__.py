"""Import for ease of use."""