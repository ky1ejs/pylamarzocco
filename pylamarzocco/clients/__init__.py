"""Initialize the client."""

from .bluetooth import LaMarzoccoBluetoothClient
from .cloud import LaMarzoccoCloudClient


__all__ = [
    "LaMarzoccoBluetoothClient",
    "LaMarzoccoCloudClient",
]
