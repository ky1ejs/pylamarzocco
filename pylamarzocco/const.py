"""Constants for La Marzocco Cloud."""

from enum import StrEnum

TOKEN_EXPIRATION = 60 * 60 * 24 * 5  # 5 days

BASE_URL = "lion.lamarzocco.io"
CUSTOMER_APP_URL = f"https://{BASE_URL}/api/customer-app"


class MachineState(StrEnum):
    """Machine states."""

    BREWING_MODE = "BrewingMode"
    STANDBY = "StandBy"


class PreExtractionMode(StrEnum):
    """Pre-extraction modes."""

    PREINFUSION = "PreInfusion"
    PREBREWING = "PreBrewing"
    DISABLED = "Disabled"


class WidgetType(StrEnum):
    """Widget types."""

    CM_MACHINE_STATUS = "CMMachineStatus"
    CM_COFFEE_BOILER = "CMCoffeeBoiler"
    CM_STEAM_BOILER_LEVEL = "CMSteamBoilerLevel"
    CM_PRE_EXTRACTION = "CMPreExtraction"
    CM_PRE_BREWING = "CMPreBrewing"
    CM_BACK_FLUSH = "CMBackFlush"
    CM_MACHINE_GROUP_STATUS = "CMMachineGroupStatus"
    CM_STEAM_BOILER_TEMPERATURE = "CMSteamBoilerTemperature"
    CM_GROUP_DOSES = "CMGroupDoses"
    CM_PRE_INFUSION_ENABLE = "CMPreInfusionEnable"
    CM_PRE_INFUSION = "CMPreInfusion"
    CM_BREW_BY_WEIGHT_DOSES = "CMBrewByWeightDoses"
    CM_CUP_WARMER = "CMCupWarmer"
    CM_HOT_WATER_DOSE = "CMHotWaterDose"
    CM_AUTO_FLUSH = "CMAutoFlush"
    CM_RINSE_FLUSH = "CMRinseFlush"
    CM_STEAM_FLUSH = "CMSteamFlush"
    CM_NO_WATER = "CMNoWater"
    G_MACHINE_STATUS = "GMachineStatus"
    G_DOSES = "GDoses"
    G_SINGLE_DOSE_MODE = "GSingleDoseMode"
    G_BARISTA_LIGHT = "GBaristaLight"
    G_HOPPER_OPENED = "GHopperOpened"
    G_MIRROR_DOSES = "GMirrorDoses"
    G_MORE_DOSE = "GMoreDose"
    G_GRIND_WITH = "GGrindWith"


class SteamTargetLevel(StrEnum):
    """Steam target levels."""

    LEVEL_1 = "Level1"
    LEVEL_2 = "Level2"
    LEVEL_3 = "Level3"


class StompMessageType(StrEnum):
    """Stomp message types."""

    CONNECT = "CONNECT"
    CONNECTED = "CONNECTED"
    DISCONNECT = "DISCONNECT"
    SUBSCRIBE = "SUBSCRIBE"
    UNSUBSCRIBE = "UNSUBSCRIBE"
    MESSAGE = "MESSAGE"


class DeviceType(StrEnum):
    """Device types."""

    MACHINE = "CoffeeMachine"
    GRINDER = "Grinder"


class ModelCode(StrEnum):
    """Model codes."""

    LINEA_MINI = "LINEAMINI"
    LINEA_MICRA = "LINEAMICRA"
    LINEA_MINI_R = "LINEAMINIR"
    GS3 = "GS3"
    GS3_MP = "GS3MP"
    GS3_AV = "GS3AV"


class FirmwareType(StrEnum):
    """Firmware types."""

    MACHINE = "Machine"
    GATEWAY = "Gateway"


class DoseIndexType(StrEnum):
    """Dose index types."""

    BY_GROUP = "ByGroup"
    BY_DOSE = "ByDose"  # TODO: Check if this is correct


class SmartStandByType(StrEnum):
    """Smart Standby types."""

    LAST_BREW = "LastBrewing"
    POWER_ON = "PowerOn"
