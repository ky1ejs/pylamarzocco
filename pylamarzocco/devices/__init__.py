"""Init the devices module."""

from ._machine import LaMarzoccoMachine
from ._thing import LaMarzoccoThing

__all__ = [
    "LaMarzoccoMachine",
    "LaMarzoccoThing",
]
