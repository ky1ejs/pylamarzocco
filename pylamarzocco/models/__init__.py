"""Models for the La Marzocco API."""

from ._authentication import AccessToken, RefreshTokenRequest, SigninTokenRequest
from ._config import (
    BackFlush,
    BaseDoseSettings,
    BrewByWeightDoses,
    BrewByWeightDoseSettings,
    CoffeeBoiler,
    DosePulsesType,
    DoseSettings,
    GroupDosesSettings,
    HotWaterDose,
    InOutTime,
    MachineStatus,
    NextStatus,
    PreBrewInfusionTime,
    PrebrewInfusionTimeLists,
    PreBrewing,
    PrebrewSettingTimes,
    PreExtraction,
    PreExtractionBase,
    PreExtractionBaseTimes,
    PreExtractionInOutTimes,
    PreExtractionPreBrewInfusionTimes,
    SecondsInOut,
    SteamBoilerLevel,
    SteamBoilerTemperature,
    ThingConfig,
    ThingDashboardConfig,
    ThingDashboardWebsocketConfig,
    ThingScale,
    ThingSettings,
)
from ._general import (
    BaseWidget,
    BaseWidgetOutput,
    CommandResponse,
    Thing,
    WebSocketDetails,
    Widget,
)
from ._schedule import (
    SchedulingSettings,
    SmartWakeUpScheduleWebsocketConfig,
    SmartWakeUpSleepSettings,
    WakeUpScheduleSettings,
)
from ._statistics import (
    CoffeeAndFlushTrend,
    CoffeeHistoryEvent,
    LastCoffee,
    LastCoffeeList,
    ThingStatistics,
)
from ._update import FirmwareSettings, FirmwareVersion, UpdateDetails

__all__ = [
    "AccessToken",
    "BackFlush",
    "BaseDoseSettings",
    "BaseWidget",
    "BaseWidgetOutput",
    "BrewByWeightDoseSettings",
    "BrewByWeightDoses",
    "CoffeeAndFlushTrend",
    "CoffeeBoiler",
    "CoffeeHistoryEvent",
    "CommandResponse",
    "ThingConfig",
    "ThingDashboardConfig",
    "ThingDashboardWebsocketConfig",
    "DosePulsesType",
    "DoseSettings",
    "FirmwareSettings",
    "FirmwareVersion",
    "GroupDosesSettings",
    "HotWaterDose",
    "InOutTime",
    "LastCoffee",
    "LastCoffeeList",
    "MachineStatus",
    "NextStatus",
    "PreBrewInfusionTime",
    "PreBrewing",
    "PreExtraction",
    "PreExtractionBase",
    "PreExtractionBaseTimes",
    "PreExtractionInOutTimes",
    "PreExtractionPreBrewInfusionTimes",
    "PrebrewInfusionTimeLists",
    "PrebrewSettingTimes",
    "RefreshTokenRequest",
    "SchedulingSettings",
    "SecondsInOut",
    "SigninTokenRequest",
    "SmartWakeUpScheduleWebsocketConfig",
    "SmartWakeUpSleepSettings",
    "SteamBoilerLevel",
    "SteamBoilerTemperature",
    "Thing",
    "ThingScale",
    "ThingSettings",
    "ThingStatistics",
    "UpdateDetails",
    "WakeUpScheduleSettings",
    "WebSocketDetails",
    "Widget",
]
