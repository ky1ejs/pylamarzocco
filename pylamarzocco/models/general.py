"""Models for general info"""

from dataclasses import dataclass, field

from mashumaro import field_options
from mashumaro.mixins.json import DataClassJSONMixin


@dataclass(kw_only=True)
class CommandResponse(DataClassJSONMixin):
    """Response for change setting endpoint"""
    id: str
    status: str
    error_code: str | None = field(
        metadata=field_options(alias="errorCode")
    )