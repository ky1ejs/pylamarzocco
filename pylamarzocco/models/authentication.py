"""Models for authentication."""

from time import time
from dataclasses import dataclass, field

from mashumaro import field_options
from mashumaro.mixins.json import DataClassJSONMixin

from pylamarzocco.const import TOKEN_EXPIRATION


@dataclass(kw_only=True)
class AccessToken(DataClassJSONMixin):
    """Token response model."""

    id: str 
    access_token: str = field(metadata=field_options(alias="accessToken"))
    refresh_token: str = field(metadata=field_options(alias="refreshToken"))
    token_type: str = field(metadata=field_options(alias="tokenType"))
    username: str
    email: str
    expires_in: float = field(default=time() + TOKEN_EXPIRATION)


@dataclass(kw_only=True)
class TokenRequest(DataClassJSONMixin):
    """Token request model."""

    username: str
    password: str