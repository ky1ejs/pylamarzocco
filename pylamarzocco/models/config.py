"""Models for device configuration."""

from __future__ import annotations
from typing import Annotated

from dataclasses import dataclass, field
from datetime import datetime, timezone

from mashumaro import field_options
from mashumaro.mixins.json import DataClassJSONMixin
from mashumaro.types import Discriminator

from pylamarzocco.models.general import CommandResponse

from pylamarzocco.const import MachineState, PreExtractionMode

@dataclass(kw_only=True)
class DeviceConfig(DataClassJSONMixin):
    """Device configuration."""

    connected: bool
    removed_widgets: list[BaseWidget] = field(
        metadata=field_options(alias="removedWidgets"), default_factory=list
    )
    connection_date: int = field(metadata=field_options(alias="connectionDate"))
    widgets: list[Widget] = field(default_factory=list)
    uuid: str
    commands: list[CommandResponse]


@dataclass(kw_only=True)
class BaseWidget(DataClassJSONMixin):
    """Base widget configuration."""

    code: str
    index: int


@dataclass(kw_only=True)
class Widget(BaseWidget):
    """Widget configuration."""

    output: list[
        Annotated[WidgetBaseOutput, Discriminator(field="code", include_subtypes=True)]
    ]

@dataclass(kw_only=True)
class WidgetBaseOutput(DataClassJSONMixin):
    """Widget configuration."""

@dataclass(kw_only=True)
class CMMachineStatus(WidgetBaseOutput):
    """Machine status configuration."""

    status: MachineState
    available_modes: list[MachineState] = field(
        metadata=field_options(alias="availableModes")
    )
    mode: MachineState
    next_state: MachineState | None = field(metadata=field_options(alias="nextState"))
    brewing_start_time: datetime | None = field(
        metadata=field_options(
            alias="brewingStartTime",
            serialize=lambda ts: datetime.fromtimestamp(ts, timezone.utc),
        ),
        default=None,
    )


@dataclass(kw_only=True)
class CMCoffeeBoiler(WidgetBaseOutput):
    """Coffee boiler configuration."""

    status: MachineState
    enabled: bool
    enabled_supported: bool = field(metadata=field_options(alias="enabledSupported"))
    target_temperature: float = field(metadata=field_options(alias="targetTemperature"))
    target_temperature_min: int = field(
        metadata=field_options(alias="targetTemperatureMin")
    )
    target_temperature_max: int = field(
        metadata=field_options(alias="targetTemperatureMax")
    )
    target_temperature_step: float = field(
        metadata=field_options(alias="targetTemperatureStep")
    )
    ready_start_time: datetime | None = field(
        metadata=field_options(
            alias="readyStartTime",
            serialize=lambda ts: datetime.fromtimestamp(ts, timezone.utc),
        ),
        default=None,
    )

@dataclass(kw_only=True)
class CMSteamBoilerLevel(WidgetBaseOutput):
    """Steam boiler level configuration."""
    status: MachineState # TODO: correct type
    enabled: bool
    enabled_supported: bool = field(metadata=field_options(alias="enabledSupported"))
    target_level: float = field(metadata=field_options(alias="targetLevel"))
    target_level_supported: bool = field(
        metadata=field_options(alias="targetLevelSupported")
    )
    ready_start_time: datetime | None = field(
        metadata=field_options(
            alias="readyStartTime",
            serialize=lambda ts: datetime.fromtimestamp(ts, timezone.utc),
        ),
        default=None,
    )

@dataclass(kw_only=True)
class PreExtractionBase(WidgetBaseOutput):
    """Pre-extraction configuration."""
    available_modes: list[PreExtractionMode] = field(
        metadata=field_options(alias="availableModes")
    )
    mode: PreExtractionMode

@dataclass(kw_only=True)
class CMPreExtraction(PreExtractionBase):
    """Pre-extraction configuration."""

    times: InOutTime

@dataclass(kw_only=True)
class CMPreBrewing(PreExtractionBase):
    """Pre-brewing configuration."""

    times: PrebrewInfusionTimeLists
    dose_index_supported: bool = field(metadata=field_options(alias="doseIndexSupported"), default=False)

@dataclass(kw_only=True)
class InOutTime(DataClassJSONMixin):
    """In and out time configuration."""

    time_in: PreExtractionInOutTimes = field(metadata=field_options(alias="In"))
    time_out: PreExtractionInOutTimes = field(metadata=field_options(alias="Out"))

@dataclass(kw_only=True)
class SecondsInOut(DataClassJSONMixin):
    """Seconds in and out configuration."""
    seconds_in: float = field(metadata=field_options(alias="In"))
    seconds_out: float = field(metadata=field_options(alias="Out"))

@dataclass(kw_only=True)
class PreBrewInfusionTime(DataClassJSONMixin):
    """Prebrew/-infusion configuration."""
    pre_infusion: float = field(metadata=field_options(alias="PreInfusion"))
    pre_brewing: float = field(metadata=field_options(alias="PreBrewing"))

@dataclass(kw_only=True)
class PreExtractionBaseTimes[T](DataClassJSONMixin):
    """Pre-extraction times configuration."""
    seconds_min: T = field(
        metadata=field_options(alias="secondsMin")
    )
    seconds_max: T = field(
        metadata=field_options(alias="secondsMax")
    )
    seconds_step: T = field(
        metadata=field_options(alias="secondsStep")
    )   


@dataclass(kw_only=True)
class PreExtractionPreBrewInfusionTimes(PreExtractionBaseTimes[PreBrewInfusionTime]):
    """Pre-extraction times configuration."""
    seconds: float

@dataclass(kw_only=True)
class PreExtractionInOutTimes(PreExtractionBaseTimes[SecondsInOut]):
    """Pre-extraction times configuration."""
    dose_index: str = field(metadata=field_options(alias="doseIndex"))


@dataclass(kw_only=True)
class PrebrewInfusionTimeLists(DataClassJSONMixin):
    """Prebrew/-infusion configuration."""
    pre_infusion: list[PreExtractionInOutTimes] = field(metadata=field_options(alias="PreInfusion"))
    pre_brewing: list[PreExtractionInOutTimes] = field(metadata=field_options(alias="PreBrewing"))

@dataclass(kw_only=True)
class CMBackFlush(WidgetBaseOutput):
    """Backflush configuration."""

    enabled: bool
    last_cleaning_start_time: datetime | None = field(
        metadata=field_options(
            alias="lastCleaningStartTime",
            serialize=lambda ts: datetime.fromtimestamp(ts, timezone.utc),
        ),
        default=None,
    )