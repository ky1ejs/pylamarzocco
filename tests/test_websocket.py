"""Tests for websocket related functionality."""

from pylamarzocco.models.config import DeviceConfig
from pylamarzocco.const import StompMessageType
from pylamarzocco.util import encode_stomp_ws_message, decode_stomp_ws_message
from .conftest import load_fixture

import json

from syrupy import SnapshotAssertion

async def test_encode_stomp_ws_message(snapshot: SnapshotAssertion) -> None:
    """Test the encode_stomp_ws_message function."""
    
    fixture = load_fixture("machine", "config_micra.json")
    device_config = DeviceConfig.from_dict(fixture)
    headers = {
        "destination": "/topic/test",
        "id": "test-id",
        "subscription": "test-subscription",
    }
    message = encode_stomp_ws_message(StompMessageType.MESSAGE, headers, json.dumps(fixture))
    assert message == snapshot

    msg_type, headers_decoded, body = decode_stomp_ws_message(message)

    assert body
    assert msg_type is StompMessageType.MESSAGE
    assert headers_decoded == headers
    assert DeviceConfig.from_json(body) == device_config

