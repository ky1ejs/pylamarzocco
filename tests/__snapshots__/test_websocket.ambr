# serializer version: 1
# name: test_websocket_message
  dict({
    'backflush_enabled': False,
    'bbw_settings': None,
    'boilers': dict({
      <BoilerType.COFFEE: 'CoffeeBoiler1'>: dict({
        'current_temperature': 81,
        'enabled': True,
        'target_temperature': 94,
      }),
      <BoilerType.STEAM: 'SteamBoiler'>: dict({
        'current_temperature': 113,
        'enabled': True,
        'target_temperature': 131,
      }),
    }),
    'brew_active': False,
    'brew_active_duration': 0,
    'dose_hot_water': 8,
    'doses': dict({
      <PhysicalKey.A: 1>: 135,
      <PhysicalKey.B: 2>: 97,
      <PhysicalKey.C: 3>: 108,
      <PhysicalKey.D: 4>: 121,
    }),
    'plumbed_in': True,
    'prebrew_configuration': dict({
      <PhysicalKey.A: 1>: dict({
        'off_time': 1,
        'on_time': 0.5,
      }),
      <PhysicalKey.B: 2>: dict({
        'off_time': 1,
        'on_time': 0.5,
      }),
      <PhysicalKey.C: 3>: dict({
        'off_time': 3.299999952316284,
        'on_time': 3.299999952316284,
      }),
      <PhysicalKey.D: 4>: dict({
        'off_time': 2,
        'on_time': 2,
      }),
    }),
    'prebrew_mode': <PrebrewMode.PREINFUSION: 'TypeB'>,
    'scale': None,
    'smart_standby': dict({
      'enabled': True,
      'minutes': 10,
      'mode': <SmartStandbyMode.LAST_BREWING: 'LastBrewing'>,
    }),
    'turned_on': True,
    'wake_up_sleep_entries': dict({
      'Os2OswX': dict({
        'days': list([
          <WeekDay.MONDAY: 'monday'>,
          <WeekDay.TUESDAY: 'tuesday'>,
          <WeekDay.WEDNESDAY: 'wednesday'>,
          <WeekDay.THURSDAY: 'thursday'>,
          <WeekDay.FRIDAY: 'friday'>,
          <WeekDay.SATURDAY: 'saturday'>,
          <WeekDay.SUNDAY: 'sunday'>,
        ]),
        'enabled': False,
        'entry_id': 'Os2OswX',
        'steam': True,
        'time_off': '24:0',
        'time_on': '24:0',
      }),
    }),
    'water_contact': True,
  })
# ---
# name: test_websocket_message.1
  dict({
    'backflush_enabled': False,
    'bbw_settings': None,
    'boilers': dict({
      <BoilerType.COFFEE: 'CoffeeBoiler1'>: dict({
        'current_temperature': 36,
        'enabled': True,
        'target_temperature': 94,
      }),
      <BoilerType.STEAM: 'SteamBoiler'>: dict({
        'current_temperature': 50,
        'enabled': True,
        'target_temperature': 131,
      }),
    }),
    'brew_active': False,
    'brew_active_duration': 0,
    'dose_hot_water': 8,
    'doses': dict({
      <PhysicalKey.A: 1>: 135,
      <PhysicalKey.B: 2>: 97,
      <PhysicalKey.C: 3>: 108,
      <PhysicalKey.D: 4>: 121,
    }),
    'plumbed_in': True,
    'prebrew_configuration': dict({
      <PhysicalKey.A: 1>: dict({
        'off_time': 1,
        'on_time': 0.5,
      }),
      <PhysicalKey.B: 2>: dict({
        'off_time': 1,
        'on_time': 0.5,
      }),
      <PhysicalKey.C: 3>: dict({
        'off_time': 3.299999952316284,
        'on_time': 3.299999952316284,
      }),
      <PhysicalKey.D: 4>: dict({
        'off_time': 2,
        'on_time': 2,
      }),
    }),
    'prebrew_mode': <PrebrewMode.PREINFUSION: 'TypeB'>,
    'scale': None,
    'smart_standby': dict({
      'enabled': True,
      'minutes': 10,
      'mode': <SmartStandbyMode.LAST_BREWING: 'LastBrewing'>,
    }),
    'turned_on': True,
    'wake_up_sleep_entries': dict({
      'Os2OswX': dict({
        'days': list([
          <WeekDay.MONDAY: 'monday'>,
          <WeekDay.TUESDAY: 'tuesday'>,
          <WeekDay.WEDNESDAY: 'wednesday'>,
          <WeekDay.THURSDAY: 'thursday'>,
          <WeekDay.FRIDAY: 'friday'>,
          <WeekDay.SATURDAY: 'saturday'>,
          <WeekDay.SUNDAY: 'sunday'>,
        ]),
        'enabled': False,
        'entry_id': 'Os2OswX',
        'steam': True,
        'time_off': '24:0',
        'time_on': '24:0',
      }),
    }),
    'water_contact': True,
  })
# ---
